be/src/storage/table_reader.cpp

```diff
@@ -14,22 +14,121 @@
 
 #include "storage/table_reader.h"
 
+#include "column/datum_convert.h"
+#include "column/datum_tuple.h"
+#include "storage/chunk_helper.h"
+#include "storage/projection_iterator.h"
+#include "storage/storage_engine.h"
+#include "storage/tablet_manager.h"
+#include "storage/tablet_reader.h"
+#include "storage/union_iterator.h"
+
 namespace starrocks {
 
-TableReader::TableReader(const TableReaderParams& params) : _params(params) {}
+TableReader::TableReader(const TableReaderParams& params) : _params(params) {
+    for (auto& partition : _params.partition_param.partitions) {
+        for (auto& index : partition.indexes) {
+            for (int64_t tablet_id : index.tablets) {
+                TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true);
+                if (tablet != nullptr) {
+                    _local_tablets.push_back(tablet);
+                }
+            }
+        }
+    }
+
+    _tablet_schema = ChunkHelper::convert_schema_to_format_v2(_local_tablets[0]->tablet_schema());
+    CHECK_EQ(_params.schema.num_fields(), _tablet_schema.num_fields());
+    for (int i = 0; i < _params.schema.num_fields(); i++) {
+        CHECK_EQ(_params.schema.field(i)->name(), _tablet_schema.field(i)->name());
+    }
+}
 
 TableReader::~TableReader() {}
 
 Status TableReader::multi_get(const Chunk& keys, const std::vector<std::string>& value_columns,
                               std::vector<bool>& found, Chunk& values) {
-    // TODO implement multi_get
+    DeferOp defer([&] { _obj_pool.clear(); });
```

Comment:
So this class is not thread-safe, only support one reader at one time?

Reply:
not share ObjectPool across multiple requests to support concurrency